File: app/lib/domain/entities/game_view.dart
import 'card.dart';
import 'table_slot.dart';

enum GamePhase { inProgress, finished }

GamePhase _phaseFromWire(String value) => value == 'finished' ? GamePhase.finished : GamePhase.inProgress;

class YouInfo {
  final String id;
  final String name;
  final List<GameCard> hand;

  const YouInfo({required this.id, required this.name, required this.hand});

  factory YouInfo.fromJson(Map<String, dynamic> json) => YouInfo(
        id: json['id'] as String,
        name: json['name'] as String,
        hand: (json['hand'] as List).map((c) => GameCard.fromJson(c as Map<String, dynamic>)).toList(),
      );
}

class OpponentInfo {
  final String id;
  final String name;
  final int cardCount;

  const OpponentInfo({required this.id, required this.name, required this.cardCount});

  factory OpponentInfo.fromJson(Map<String, dynamic> json) => OpponentInfo(
        id: json['id'] as String,
        name: json['name'] as String,
        cardCount: json['cardCount'] as int,
      );
}

/// Mirrors the server's redacted per-player projection (`GameView` on the
/// backend) — the client only ever sees its own hand, never the opponent's.
class GameView {
  final String gameId;
  final GamePhase phase;
  final YouInfo you;
  final OpponentInfo opponent;
  final int deckCount;
  final GameCard trumpCard;
  final Suit trumpSuit;
  final List<TableSlot> table;
  final String attackerId;
  final String defenderId;
  final List<String> winnerOrder;
  final String? loserId;

  const GameView({
    required this.gameId,
    required this.phase,
    required this.you,
    required this.opponent,
    required this.deckCount,
    required this.trumpCard,
    required this.trumpSuit,
    required this.table,
    required this.attackerId,
    required this.defenderId,
    required this.winnerOrder,
    this.loserId,
  });

  bool get isYouAttacker => you.id == attackerId;
  bool get isYouDefender => you.id == defenderId;

  factory GameView.fromJson(Map<String, dynamic> json) => GameView(
        gameId: json['gameId'] as String,
        phase: _phaseFromWire(json['phase'] as String),
        you: YouInfo.fromJson(json['you'] as Map<String, dynamic>),
        opponent: OpponentInfo.fromJson(json['opponent'] as Map<String, dynamic>),
        deckCount: json['deckCount'] as int,
        trumpCard: GameCard.fromJson(json['trumpCard'] as Map<String, dynamic>),
        trumpSuit: suitFromWire(json['trumpSuit'] as String),
        table: (json['table'] as List).map((s) => TableSlot.fromJson(s as Map<String, dynamic>)).toList(),
        attackerId: json['attackerId'] as String,
        defenderId: json['defenderId'] as String,
        winnerOrder: (json['winnerOrder'] as List).cast<String>(),
        loserId: json['loserId'] as String?,
      );
}

File: server/src/infrastructure/game-view.ts
import { Card } from "../domain/card.js";
import { GameState, otherPlayerId } from "../domain/game-state.js";

/**
 * A player should never receive the opponent's actual hand over the wire,
 * only how many cards they hold. This projects the shared GameState into
 * a per-recipient DTO safe to send over the socket.
 */
export interface GameView {
  gameId: string;
  phase: GameState["phase"];
  you: { id: string; name: string; hand: Card[] };
  opponent: { id: string; name: string; cardCount: number };
  deckCount: number;
  discardCount: number;
  trumpCard: Card;
  trumpSuit: Card["suit"];
  table: { attack: Card; defense?: Card }[];
  attackerId: string;
  defenderId: string;
  winnerOrder: string[];
  loserId?: string;
}

export function toGameView(state: GameState, forPlayerId: string): GameView {
  const you = state.players.find((p) => p.id === forPlayerId)!;
  const opponentId = otherPlayerId(state, forPlayerId);
  const opponent = state.players.find((p) => p.id === opponentId)!;

  return {
    gameId: state.id,
    phase: state.phase,
    you: { id: you.id, name: you.name, hand: you.hand },
    opponent: { id: opponent.id, name: opponent.name, cardCount: opponent.hand.length },
    deckCount: state.deck.length,
    discardCount: state.discardCount,
    trumpCard: state.trumpCard,
    trumpSuit: state.trumpSuit,
    table: state.table,
    attackerId: state.attackerId,
    defenderId: state.defenderId,
    winnerOrder: state.winnerOrder,
    loserId: state.loserId,
  };
}
